File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) { this.val = val; this.next = null; }
 * }
 */

class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }
        ListNode t = head;
        ListNode h = head;
        boolean flag = false;
        while (h != null && h.next != null) {
            t = t.next;
            h = h.next.next;
            if (h == t) {
                flag = true;
                break;
            }
        }
        if (flag) {
            t = head;
            while (t != h) {
                t = t.next;
                h = h.next;
            }
            return t;
        }
        return null;
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast=head,*slow=head;
        while(fast){
            fast=fast->next;
            if(fast)fast=fast->next;
            else return NULL;
            slow=slow->next;
            if(fast==slow)break;
        }
        if(!fast)return NULL;
        ListNode* entry=head;
        while(entry!=slow){
            entry=entry->next;
            slow=slow->next;
        } 
        return entry;
    }
};